proxylib/proxylib.go

```diff
@@ -0,0 +1,227 @@
+		return C.FILTER_POLICY_DROP
+	}
+
+	mutex.Lock()
+	connections[connectionId] = connection
+	mutex.Unlock()
+
+	return C.FILTER_OK
+}
+
+// Each connection is assumed to be called from a single thread, so accessing connection metadata
+// does not need protection.
+//
+// OnData gets all the unparsed data the datapath has received so far. The data is provided to the parser
+// associated with the connection, and the parser is expected to find if the data frame contains enough data
+// to make a PASS/DROP decision for the whole data frame. Note that the whole data frame need not be received,
+// if the decision including the length of the data frame in bytes can be determined based on the beginning of
+// the data frame only (e.g., headers including the length of the data frame). The parser returns a decision
+// with the number of bytes on which the decision applies. If more data is available, then the parser will be
+// called again with the remaining data. Parser needs to return MORE if a decision can't be made with
+// the available data, including the minimum number of additional bytes that is needed before the parser is
+// called again.
+//
+// The parser can also inject at arbitrary points in the data stream. This is indecated by an INJECT operation
+// with the number of bytes to be injected. The actual bytes to be injected are provided via an Inject()
+// callback prior to returning the INJECT operation. The Inject() callback operates on a limited size buffer
+// provided by the datapath, and multiple INJECT operations may be needed to inject large amounts of data.
+// Since we get the data on one direction at a time, any frames to be injected in the reverse direction
+// are placed in the reverse direction buffer, from where the datapath injects the data before calling
+// us again for the reverse direction input.
+//
+//export OnData
+func OnData(connectionId uint64, reply, endStream bool, data *[][]byte, filterOps *[]C.FilterOp) C.FilterResult {
+	// Find the connection
+	mutex.RLock()
+	connection, ok := connections[connectionId]
+	mutex.RUnlock()
+	if !ok {
+		return C.FILTER_UNKNOWN_CONNECTION
+	}
+
+	unit := 0
+	offset := 0
+
+	// Loop until `filterOps` becomes full, or parser is done with the data.
+	for len(*filterOps) < cap(*filterOps) {
+		op, bytes := connection.Parser.OnData(reply, endStream, (*data)[unit:], offset)
+		if op == NOP {
+			break // No operations after NOP
+		}
+		if bytes == 0 {
+			return C.FILTER_PARSER_ERROR
+		}
+		*filterOps = append(*filterOps, C.FilterOp{C.uint64_t(op), C.int64_t(bytes)})
+
+		if op == MORE {
+			// Need more data before can parse ahead.
+			// Parser will see the unused data again in the next call, which will take place
+			// after there are at least 'bytes' of additional data to parse.
+			break
+		}
+
+		if op == PASS || op == DROP {
+			// Skip bytes in input, or exhaust the input.
+			for bytes > 0 && unit < len(*data) {
```

Comment:
Non-blocking: Would be great to isolate such functionality in a function and add a unit test.